# Chapter 9 



In [1]:
!pip install -q numpy 

In [2]:
def array_properties(a, label='a'):
    print(label, '= ')
    print(a)
    print("dim = ", a.ndim)
    print("shape = ", a.shape)
    print("datatype = ", a.dtype)
    print("size = ", a.size)
    print()

# Linear Algebra 2 


## Creating a matrix representation of system of linear equation



$$
A . x = b
$$

$$
x = \frac {b} {A}
$$

$$
x = A^{-1} . b
$$

$$
3x_1 -3x_2 - 4x_3 = 1 \\
x_1 - x_2 - x_3 = 2 \\
2x_1 - x_2 + 4x_3 = 3	
$$



$$
A=\left(
        \begin{matrix} 
            3 & -3 & -4 \\ 
            1 & -1 & -1 \\ 
            2 & -1 & 4 \\ 
        \end{matrix}
    \right)
$$

$$
x=\left(
        \begin{matrix} 
            x_1 \\ 
            x_2 \\ 
            x_3 \\ 
        \end{matrix}
    \right)
$$

$$
b=\left(
        \begin{matrix} 
            1 \\ 
            2 \\ 
            3 \\ 
        \end{matrix}
    \right)
$$

Solve for the values of $ x $.

In [3]:
import numpy as np

A = np.array([
    [3., -3., -4.],
    [1, -1, -1],
    [2, -1, 4],
])

array_properties(A, 'A')


A = 
[[ 3. -3. -4.]
 [ 1. -1. -1.]
 [ 2. -1.  4.]]
dim =  2
shape =  (3, 3)
datatype =  float64
size =  9



In [4]:
import numpy as np

b = np.array([
    [1.],
    [2],
    [3],
])

array_properties(b, 'b')


b = 
[[1.]
 [2.]
 [3.]]
dim =  2
shape =  (3, 1)
datatype =  float64
size =  3



In [5]:
import numpy as np

x = np.linalg.solve(A,b)

array_properties(x, 'x')

x = 
[[-24.]
 [-31.]
 [  5.]]
dim =  2
shape =  (3, 1)
datatype =  float64
size =  3



In [6]:
import numpy as np

A_inv = np.linalg.inv(A)

x = A_inv @ b

array_properties(A_inv, 'A_inv')
array_properties(x, 'x')


A_inv = 
[[  5. -16.   1.]
 [  6. -20.   1.]
 [ -1.   3.   0.]]
dim =  2
shape =  (3, 3)
datatype =  float64
size =  9

x = 
[[-24.]
 [-31.]
 [  5.]]
dim =  2
shape =  (3, 1)
datatype =  float64
size =  3



$$

x_1 = \frac {D_1} {D}

$$

In [7]:
D = np.linalg.det(A)
print('D = ', D)


D =  -0.9999999999999998


In [8]:
A1 = A.copy()


A1[:,0] = b[:,0]
array_properties(A1, 'A1')

D1 = np.linalg.det(A1)
print('D1= ', D1)

x1 = D1 / D
print('x1 = ', x1)

A1 = 
[[ 1. -3. -4.]
 [ 2. -1. -1.]
 [ 3. -1.  4.]]
dim =  2
shape =  (3, 3)
datatype =  float64
size =  9

D1=  24.000000000000014
x1 =  -24.00000000000002


In [9]:
# xx = []

for i in range(3):
    AA = A.copy()

    AA[:,i] = b[:,0]
    # array_properties(AA, f'A{i+1}')

    DD = np.linalg.det(AA)
    print(f'D{i+1}= ', DD)

    xx = DD / D
    print(f'x{i+1} = ', xx)


D1=  24.000000000000014
x1 =  -24.00000000000002
D2=  31.0
x2 =  -31.000000000000007
D3=  -5.000000000000001
x3 =  5.000000000000002


## Class Activity

1. Create an array with three concentric rectangle of values -1, 0, 1.

$$ 
\begin{bmatrix}
    -1 & -1 & -1 & -1 & -1 \\  
    -1 & 0 & 0 & 0 & -1 \\  
    -1 & 0 & 1 & 0 & -1 \\  
    -1 & 0 & 0 & 0 & -1 \\  
    -1 & -1 & -1 & -1 & -1 \\  
\end{bmatrix}
$$


1. What is the shape.
   1. 5x5
2. What is the dimension of the array.
   1. 2
3. What is the index of 1. 
   1. [2,2]

## Convert an Array to Triangular Matrix


$$ 
\begin{bmatrix}
    a_{11} & a_{12} & a_{13} & a_{14}  \\  
    a_{21} & a_{22} & a_{23} & a_{24}  \\  
    a_{31} & a_{32} & a_{33} & a_{34}  \\  
    a_{41} & a_{42} & a_{43} & a_{44}  \\  
\end{bmatrix}
$$


$$ 
\begin{bmatrix}
    a_{11} & a_{12} & a_{13} & a_{14}  \\  
    0 & a_{22} & a_{23} & a_{24}  \\  
    0 & 0 & a_{33} & a_{34}  \\  
    0 & 0 & 0 & a_{44}  \\  
\end{bmatrix}
$$


$$

R_i = R_i - R_p \times \frac {R_{i,p}} {R_{p,p}}

$$

where $p$ is the pivot
$$
p = 1, 2, ... m-1 \\ 
i = p+1, p+2, ... m
$$


$$
if \quad R_{p,p} = 0 \\  
interchange \quad R_p \quad with \quad non-zero \quad R_{i,p} row

$$

In [19]:
# using row equivalent reduction method
# convert a matrix to a triangular form


import numpy as np

# create matrix A
A = np.array([
    [3., -3., -4.],
    [1, -1, -1],
    [2, -1, 4],
])

# create matrix b
b = np.array([1,2,3]).reshape(3,1)

# create auguemented matrix [A|b]
M = np.hstack([A,b])

print('M =', M)

# 
nrows,ncols = M.shape

# reduce the column 1 and row 2 element to zero
#  
M[1,:] = M[1] - M[0]*M[1,0]/M[0,0]
print('M =', M)

# reduce the column 1 and row 3 element to zero
#  
M[2] = M[2] - M[0]*M[2,0]/M[0,0]
print('M =', M)

# check if pivot M[1,1] = 0, then interchange with a 
# non-zero below row
if M[1,1] == 0:
    # interchange row with non-zero below row
    for i in range(2,nrows):
        if M[i,1] != 0:
            # interchange now and break
            temp = M[1].copy()
            M[1] = M[i]
            M[i] = temp
            break
else:
    # reduce column 2 row 3 to zero
    M[2] = M[2] - M[1]*M[2,1]/M[1,1]

print('M =', M)



M = [[ 3. -3. -4.  1.]
 [ 1. -1. -1.  2.]
 [ 2. -1.  4.  3.]]
M = [[ 3.         -3.         -4.          1.        ]
 [ 0.          0.          0.33333333  1.66666667]
 [ 2.         -1.          4.          3.        ]]
M = [[ 3.         -3.         -4.          1.        ]
 [ 0.          0.          0.33333333  1.66666667]
 [ 0.          1.          6.66666667  2.33333333]]
M = [[ 3.         -3.         -4.          1.        ]
 [ 0.          1.          6.66666667  2.33333333]
 [ 0.          0.          0.33333333  1.66666667]]


In [23]:
# using index

# using row equivalent reduction method
# convert a matrix to a triangular form


import numpy as np

# create matrix A
A = np.array([
    [3., -3., -4.],
    [1, -1, -1],
    [2, -1, 4],
])

# create matrix b
b = np.array([1,2,3]).reshape(3,1)

# create auguemented matrix [A|b]
M = np.hstack([A,b])

print('M =', M)

# 
nrows,ncols = M.shape
# count row interchange
k = 0

for p in range(nrows-1):
    # check if pivot M[1,1] = 0, then interchange with a 
    # non-zero below row
    if M[p,p] == 0:
        # interchange row with non-zero below row
        for j in range(p+1,nrows):
            if M[j,p] != 0:
                # interchange now and break
                temp = M[p].copy()
                M[p] = M[j]
                M[j] = temp
                k += 1
                print('k =', k)
                print('M =', M)
                break
        
        for i in range(p+1, nrows):            
            # reduce rows
            M[i] = M[i] - M[p]*M[i,p]/M[p,p]
            print('M =', M)
    else:        
        for i in range(p+1, nrows):            
            # reduce rows
            M[i] = M[i] - M[p]*M[i,p]/M[p,p]
            print('M =', M)




M = [[ 3. -3. -4.  1.]
 [ 1. -1. -1.  2.]
 [ 2. -1.  4.  3.]]
M = [[ 3.         -3.         -4.          1.        ]
 [ 0.          0.          0.33333333  1.66666667]
 [ 2.         -1.          4.          3.        ]]
M = [[ 3.         -3.         -4.          1.        ]
 [ 0.          0.          0.33333333  1.66666667]
 [ 0.          1.          6.66666667  2.33333333]]
k = 1
M = [[ 3.         -3.         -4.          1.        ]
 [ 0.          1.          6.66666667  2.33333333]
 [ 0.          0.          0.33333333  1.66666667]]
M = [[ 3.         -3.         -4.          1.        ]
 [ 0.          1.          6.66666667  2.33333333]
 [ 0.          0.          0.33333333  1.66666667]]


$$
3x_1 -3x_2 - 4x_3 = 1 \\
x_1 - x_2 - x_3 = 2 \\
2x_1 - x_2 + 4x_3 = 3	
$$


$$
3x_1 -3x_2 - 4x_3 = 1 \\
  x_2 - 6.67x_3 = 2.33 \\
  + 0.33x_3 = 1.67	
$$



## Class Activity

1. Convert the Row equivalent triangular operation to a function with name to_triangular.
2. Call the function to solve our system of linear equations.

<code>  

    nrows,ncols = M.shape
    # count row interchange
    k = 0

    for p in range(nrows-1):
        # check if pivot M[1,1] = 0, then interchange with a 
        # non-zero below row
        if M[p,p] == 0:
            # interchange row with non-zero below row
            for j in range(p+1,nrows):
                if M[j,p] != 0:
                    # interchange now and break
                    temp = M[p].copy()
                    M[p] = M[j]
                    M[j] = temp
                    k += 1
                    print('k =', k)
                    print('M =', M)
                    break
            
            for i in range(p+1, nrows):            
                # reduce rows
                M[i] = M[i] - M[p]*M[i,p]/M[p,p]
                print('M =', M)
        else:        
            for i in range(p+1, nrows):            
                # reduce rows
                M[i] = M[i] - M[p]*M[i,p]/M[p,p]
                print('M =', M)

</code>

In [44]:
def to_triangular(a):    
    M = a.copy()
    # 
    nrows,ncols = M.shape
    # count row interchange
    k = 0

    for p in range(nrows-1):
        # check if pivot M[1,1] = 0, then interchange with a 
        # non-zero below row
        if M[p,p] == 0:
            # interchange row with non-zero below row
            for j in range(p+1,nrows):
                if M[j,p] != 0:
                    # interchange now and break
                    temp = M[p].copy()
                    M[p] = M[j]
                    M[j] = temp
                    k += 1
                    # print('k =', k)
                    # print('M =', M)
                    break
            
            for i in range(p+1, nrows):            
                # reduce rows
                M[i] = M[i] - M[p]*M[i,p]/M[p,p]
                # print('M =', M)
        else:        
            for i in range(p+1, nrows):            
                # reduce rows
                M[i] = M[i] - M[p]*M[i,p]/M[p,p]
                # print('M =', M)

    return M, k




In [45]:
# using index

# using row equivalent reduction method
# convert a matrix to a triangular form


import numpy as np

# create matrix A
A = np.array([
    [3., -3., -4.],
    [1, -1, -1],
    [2, -1, 4],
])

# create matrix b
b = np.array([1,2,3]).reshape(3,1)

# create auguemented matrix [A|b]
M = np.hstack([A,b])

print('M =', M)

Mt, k = to_triangular(M)

print('Mt =', Mt)
print('M =', M)




M = [[ 3. -3. -4.  1.]
 [ 1. -1. -1.  2.]
 [ 2. -1.  4.  3.]]
Mt = [[ 3.         -3.         -4.          1.        ]
 [ 0.          1.          6.66666667  2.33333333]
 [ 0.          0.          0.33333333  1.66666667]]
M = [[ 3. -3. -4.  1.]
 [ 1. -1. -1.  2.]
 [ 2. -1.  4.  3.]]


In [46]:
import numpy as np

# convert system of linear equations to triangular matrix

# create matrix A
A = np.random.randint(0, 10,size=(5,5)).astype('float32')
print('A=', A)

# create matrix b
b = np.random.randint(0, 10,size=(5,1)).astype('float32')
print('b =', b)

# create auguemented matrix [A|b]
M = np.hstack([A,b])

print('M =', M)

Mt, k = to_triangular(M)

print('Mt =', Mt)




A= [[7. 1. 8. 1. 4.]
 [5. 3. 8. 5. 4.]
 [5. 5. 6. 3. 5.]
 [6. 2. 3. 8. 2.]
 [8. 1. 2. 3. 6.]]
b = [[1.]
 [3.]
 [0.]
 [0.]
 [9.]]
M = [[7. 1. 8. 1. 4. 1.]
 [5. 3. 8. 5. 4. 3.]
 [5. 5. 6. 3. 5. 0.]
 [6. 2. 3. 8. 2. 0.]
 [8. 1. 2. 3. 6. 9.]]
Mt = [[ 7.         1.         8.         1.         4.         1.       ]
 [ 0.         2.2857141  2.2857141  4.285714   1.1428571  2.2857141]
 [ 0.         0.        -4.        -5.75       0.        -5.       ]
 [ 0.         0.         0.        12.1875    -2.         4.25     ]
 [ 0.         0.         0.         0.         3.5       12.5      ]]


## Convert an Array to Diagonal Array


$$ 
\begin{bmatrix}
    a_{11} & a_{12} & a_{13} & a_{14}  \\  
    a_{21} & a_{22} & a_{23} & a_{24}  \\  
    a_{31} & a_{32} & a_{33} & a_{34}  \\  
    a_{41} & a_{42} & a_{43} & a_{44}  \\  
\end{bmatrix}
$$


$$ 
\begin{bmatrix}
    a_{11} & a_{12} & a_{13} & a_{14}  \\  
    0 & a_{22} & a_{23} & a_{24}  \\  
    0 & 0 & a_{33} & a_{34}  \\  
    0 & 0 & 0 & a_{44}  \\  
\end{bmatrix}
$$

$$ 
\begin{bmatrix}
    a_{11} & 0 & 0 & 0  \\  
    0 & a_{22} & 0 & 0  \\  
    0 & 0 & a_{33} & 0  \\  
    0 & 0 & 0 & a_{44}  \\  
\end{bmatrix}
$$


$$
R_i = R_i - R_p \times \frac {R_{i,p}} {R_{p,p}}
$$
$$
p = m, m-1, ..., 2 \\ 
i = p-1, p-2, ..., 1 \\ 
or  \\ 
i = 1, 2, ..., p-1 \\ 
$$

In [64]:
def print_array(M):

    nrows,ncols = M.shape

    print()
    for i in range(nrows):
        for j in range(ncols):
            print(f'{M[i,j]:.2f}', end='\t')
        print()


In [65]:
def to_diagonal(Mt):
    M = Mt.copy()
    # 
    nrows,ncols = M.shape

    # diagonal
    for p in range(nrows-1, 0,-1):
        for i in range(0, p):            
            # reduce rows
            M[i] = M[i] - M[p]*M[i,p]/M[p,p]
            # print('M =', M)

    return M


In [82]:
# using row equivalent reduction method
# convert a matrix to a triangular form

import numpy as np

# create matrix A
A = np.array([
    [3., -3., -4.],
    [1, -1, -1],
    [2, -1, 4],
])

# create matrix b
b = np.array([1,2,3]).reshape(3,1)

# create auguemented matrix [A|b]
M = np.hstack([A,b])

# print('M =', M)

# triangular
Mt, k = to_triangular(M)
print_array(Mt)

# diagonal
Md = to_diagonal(Mt)
print_array(Md)



3.00	-3.00	-4.00	1.00	
0.00	1.00	6.67	2.33	
0.00	0.00	0.33	1.67	

3.00	0.00	0.00	-72.00	
0.00	1.00	0.00	-31.00	
0.00	0.00	0.33	1.67	


## Convert to Identity Matrix 

From the identity matrix the b column is transformed to the solution of the linear equations


$$ 
\begin{bmatrix}
    a_{11} & a_{12} & a_{13} & a_{14} & b_{1}  \\  
    a_{21} & a_{22} & a_{23} & a_{24} & b_{2}  \\  
    a_{31} & a_{32} & a_{33} & a_{34} & b_{3}  \\  
    a_{41} & a_{42} & a_{43} & a_{44} & b_{4}  \\  
\end{bmatrix}
$$


$$ 
\begin{bmatrix}
    a_{11} & a_{12} & a_{13} & a_{14}  & b_{1} \\  
    0 & a_{22} & a_{23} & a_{24}  & b_{2} \\  
    0 & 0 & a_{33} & a_{34}  & b_{3} \\  
    0 & 0 & 0 & a_{44}  & b_{4} \\  
\end{bmatrix}
$$

$$ 
\begin{bmatrix}
    a_{11} & 0 & 0 & 0  & b_{1} \\  
    0 & a_{22} & 0 & 0  & b_{2} \\  
    0 & 0 & a_{33} & 0  & b_{3} \\  
    0 & 0 & 0 & a_{44}  & b_{4} \\  
\end{bmatrix}
$$

$$ 
\begin{bmatrix}
    1 & 0 & 0 & 0  & x_{1} \\  
    0 & 1 & 0 & 0  & x_{2}  \\  
    0 & 0 & 1 & 0  & x_{3}  \\  
    0 & 0 & 0 & 1  & x_{4} \\  
\end{bmatrix}
$$


In [77]:
def to_identity(Md):
    # convert to identity
    Mi = Md.copy()
    nrows,ncols = Mi.shape

    for i in range(nrows):
        Mi[i] = Mi[i] / Mi[i,i]

    # print('Mi=', Mi)
    # print_array(Mi)

    return Mi



In [81]:
# using row equivalent reduction method
# convert a matrix to a triangular form

import numpy as np

# create matrix A
A = np.array([
    [3., -3., -4.],
    [1, -1, -1],
    [2, -1, 4],
])

# create matrix b
b = np.array([1,2,3]).reshape(3,1)

# create auguemented matrix [A|b]
M = np.hstack([A,b])

# print('M =', M)

# triangular
Mt, k = to_triangular(M)
print_array(Mt)

# diagonal
Md = to_diagonal(Mt)
print_array(Md)

# convert to identity
Mi = to_identity(Md)
print_array(Mi)

# get x the solution
x = Mi[:,-1]

print('\nx =', x)



3.00	-3.00	-4.00	1.00	
0.00	1.00	6.67	2.33	
0.00	0.00	0.33	1.67	

3.00	0.00	0.00	-72.00	
0.00	1.00	0.00	-31.00	
0.00	0.00	0.33	1.67	

1.00	0.00	0.00	-24.00	
0.00	1.00	0.00	-31.00	
0.00	0.00	1.00	5.00	

x = [-24. -31.   5.]


## Compute Determinant from Triangular

In [86]:
def det(Mt, k):
    # determinant
    D = 1
    for p in range(nrows):
        D *= Mt[p,p]

    # change sign by the number of rows interchaged
    return (-1)**k * D

In [88]:
# using row equivalent reduction method
# convert a matrix to a triangular form

import numpy as np

# create matrix A
A = np.array([
    [3., -3., -4.],
    [1, -1, -1],
    [2, -1, 4],
])

# create matrix b
b = np.array([1,2,3]).reshape(3,1)

# create auguemented matrix [A|b]
M = np.hstack([A,b])

# print('M =', M)

# triangular
Mt, k = to_triangular(M)
print_array(Mt)

# determinant
D = det(Mt, k)
print('\nDeterminant=', D)



3.00	-3.00	-4.00	1.00	
0.00	1.00	6.67	2.33	
0.00	0.00	0.33	1.67	

Determinant= -0.9999999999999998


## Inverse of Matrix A


In [92]:
# using row equivalent reduction method
# convert a matrix to a triangular form

import numpy as np

# create matrix A
A = np.array([
    [3., -3., -4.],
    [1, -1, -1],
    [2, -1, 4],
])

# create matrix b
b = np.array([1,2,3]).reshape(3,1)

# create indentity
nrows, ncols = A.shape
I = np.identity(nrows)

# create auguemented matrix [A|b]
M = np.hstack([A,b,I])

print('M =', M)

# triangular
Mt, k = to_triangular(M)
print_array(Mt)

# determinant
D = det(Mt, k)
print('\nDeterminant=', D)

# diagonal
Md = to_diagonal(Mt)
print_array(Md)

# convert to identity
Mi = to_identity(Md)
print_array(Mi)

# get x the solution
x = Mi[:,nrows]

print('\nx =', x)

# inverse
invA = Mi[:,nrows+1:]
print_array(invA)

M = [[ 3. -3. -4.  1.  1.  0.  0.]
 [ 1. -1. -1.  2.  0.  1.  0.]
 [ 2. -1.  4.  3.  0.  0.  1.]]

3.00	-3.00	-4.00	1.00	1.00	0.00	0.00	
0.00	1.00	6.67	2.33	-0.67	0.00	1.00	
0.00	0.00	0.33	1.67	-0.33	1.00	0.00	

Determinant= -0.9999999999999998

3.00	0.00	0.00	-72.00	15.00	-48.00	3.00	
0.00	1.00	0.00	-31.00	6.00	-20.00	1.00	
0.00	0.00	0.33	1.67	-0.33	1.00	0.00	

1.00	0.00	0.00	-24.00	5.00	-16.00	1.00	
0.00	1.00	0.00	-31.00	6.00	-20.00	1.00	
0.00	0.00	1.00	5.00	-1.00	3.00	0.00	

x = [-24. -31.   5.]

5.00	-16.00	1.00	
6.00	-20.00	1.00	
-1.00	3.00	0.00	
